src/libs/history.c

```diff
@@ -169,6 +175,64 @@ static GtkWidget *_lib_history_create_button(dt_lib_module_t *self, int num, con
   return widget;
 }
 
+static GList *_duplicate_history(GList *hist)
+{
+  GList *result = NULL;
+
+  GList *h = g_list_first(hist);
+  while(h)
+  {
+    const dt_dev_history_item_t *old = (dt_dev_history_item_t *)(h->data);
+
+    dt_dev_history_item_t *new = (dt_dev_history_item_t *)malloc(sizeof(dt_dev_history_item_t));
+
+    memcpy(new, old, sizeof(dt_dev_history_item_t));
+
+    new->params = malloc(old->module->params_size);
+    new->blend_params = malloc(sizeof(dt_develop_blend_params_t));
+
+    memcpy(new->params, old->params, old->module->params_size);
+    memcpy(new->blend_params, old->blend_params, sizeof(dt_develop_blend_params_t));
+
+    result = g_list_append(result, new);
+
+    h = g_list_next(h);
+  }
+  return result;
+}
+
+static void pop_undo(dt_lib_module_t *self, dt_undo_type_t type, dt_undo_data_t *data)
+{
+  if(type == DT_UNDO_HISTORY)
+  {
+    dt_lib_history_t *d = (dt_lib_history_t *)self->data;
+    dt_undo_history_t *hist = (dt_undo_history_t *)data;
+
+    const GList *current_params = darktable.develop->history;
+    const int end = darktable.develop->history_end;
+
+    darktable.develop->history = hist->snapshot;
+    darktable.develop->history_end = hist->end;
+
+    hist->snapshot = current_params;
```

Comment:
[ 39%] Building C object src/libs/CMakeFiles/history.dir/history.c.o
/home/lebedevri/darktable/src/libs/history.c: In function ‘pop_undo’:
/home/lebedevri/darktable/src/libs/history.c:217:20: error: assignment discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
     hist->snapshot = current_params;
                    ^
/home/lebedevri/darktable/src/libs/history.c: In function ‘_history_undo_data_free’:
/home/lebedevri/darktable/src/libs/history.c:232:30: error: passing argument 2 of ‘g_list_free_full’ from incompatible pointer type [-Werror=incompatible-pointer-types]
   g_list_free_full(snapshot, dt_dev_free_history_item);
                              ^
In file included from /usr/include/glib-2.0/glib/ghash.h:33:0,
                 from /usr/include/glib-2.0/glib.h:50,
                 from /home/lebedevri/darktable/src/common/dtpthread.h:29,
                 from /home/lebedevri/darktable/src/common/darktable.h:37,
                 from /home/lebedevri/darktable/src/libs/history.c:19:
/usr/include/glib-2.0/glib/glist.h:56:10: note: expected ‘GDestroyNotify {aka void (*)(void *)}’ but argument is of type ‘void (*)(dt_dev_history_item_t *) {aka void (*)(struct dt_dev_history_item_t *)}’
 void     g_list_free_full               (GList            *list,
          ^
/home/lebedevri/darktable/src/libs/history.c: In function ‘_lib_history_change_callback’:
/home/lebedevri/darktable/src/libs/history.c:259:38: error: passing argument 2 of ‘dt_undo_record’ from incompatible pointer type [-Werror=incompatible-pointer-types]
       dt_undo_record(darktable.undo, self, DT_UNDO_HISTORY, (dt_undo_data_t *)hist, &pop_undo, _history_undo_data_free);
                                      ^
In file included from /home/lebedevri/darktable/src/libs/history.c:26:0:
/home/lebedevri/darktable/src/views/undo.h:61:6: note: expected ‘dt_view_t * {aka struct dt_view_t *}’ but argument is of type ‘dt_lib_module_t * {aka struct dt_lib_module_t *}’
 void dt_undo_record(dt_undo_t *self, dt_view_t *view, dt_undo_type_t type, dt_undo_data_t *data,
      ^
/home/lebedevri/darktable/src/libs/history.c:259:85: error: passing argument 5 of ‘dt_undo_record’ from incompatible pointer type [-Werror=incompatible-pointer-types]
       dt_undo_record(darktable.undo, self, DT_UNDO_HISTORY, (dt_undo_data_t *)hist, &pop_undo, _history_undo_data_free);
                                                                                     ^
In file included from /home/lebedevri/darktable/src/libs/history.c:26:0:
/home/lebedevri/darktable/src/views/undo.h:61:6: note: expected ‘void (*)(dt_view_t *, dt_undo_type_t,  void **) {aka void (*)(struct dt_view_t *, enum dt_undo_type_t,  void **)}’ but argument is of type ‘void (*)(dt_lib_module_t *, dt_undo_type_t,  void **) {aka void (*)(struct dt_lib_module_t *, enum dt_undo_type_t,  void **)}’
 void dt_undo_record(dt_undo_t *self, dt_view_t *view, dt_undo_type_t type, dt_undo_data_t *data,
      ^
cc1: all warnings being treated as errors
src/libs/CMakeFiles/history.dir/build.make:62: recipe for target 'src/libs/CMakeFiles/history.dir/history.c.o' failed
make[2]: *** [src/libs/CMakeFiles/history.dir/history.c.o] Error 1
make[2]: Target 'src/libs/CMakeFiles/history.dir/build' not remade because of errors.
CMakeFiles/Makefile2:2295: recipe for target 'src/libs/CMakeFiles/history.dir/all' failed
make[1]: *** [src/libs/CMakeFiles/history.dir/all] Error 2